STEP 4-1 기본 구현

hwj0623 opened the pull request:
[ 1st commit ]
- 레포지토리 포크 후 테스트 커밋 

[ 2nd commit ]
- 2차원배열을 활용한 사다리게임 기본 형태 구현

[ 3rd commit ]
- 메소드명 변경: 2진 랜덤 숫자 생성 함수, 숫자 입력 함수
- stride 함수 적용.
- 입력 유효범위에 대한 변경

hwj0623 committed:
STEP 0 : initial test commit from forked repository

hwj0623 committed:
STEP 4-1 : 사다리게임 프로그램 구현

hwj0623 committed:
STEP 4-1 : 변수명 수정, stride 적용

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -5,6 +5,89 @@
+    guard let n: String = readLine()
+        else{
+            print("유효한 값을 입력하세요")
+            return (-1, -1)
```
입력을 담당하는 함수에서 잘못된 값이라고 해서 에러 문구를 출력하는 것은 좋은 선택이 아닙니다.
- 입력하는 함수는 결과를 return 해주고, 호출했던 상위 모듈에서 에러 문구를 출력하거나 흐름을 제어해서 다시 호출하기만 하면 되는거죠. 
- `return (-1, -1)` 형태로 값을 전달하는 곳에 에러를 의미하는 값을 전달하는 것도 좋지 않습니다. 차라리 Result 에 해당하는 값을 별도로 주는 것이 안전합니다. 값에 대한 범위로 예외처리를 하지는 마세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -5,6 +5,89 @@
+        let resultLadder: [[String]] = buildLadder (ladderInt2dArray : initialLadder)
+        printLadder(ladderString2dArray: resultLadder)
+    }else{
+        print("유효한 입력값을 넣어주세요 (인원, 높이 각각 2이상) \n. 양의 실수값 입력시 정수로 내림 처리됩니다.")
```
inputPairNumber()에서 실패한 경우는 실제로 여기 print() 문구와 input 에러 문구가 중복해서 나오고 있네요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -5,6 +5,89 @@
+}
+/// 2차원 사다리 문자열 생성 함수
+func  buildLadder (ladderInt2dArray : [[String]]) -> [[String]] {
+    var resultLadder = ladderInt2dArray
```
ladderInt2dArray 변수 이름에 Int, Array, String 같은 타입을 붙이지 마세요. 
대신 어떤 역할을 하는지, 어떤 의도로 사용하는 것인지 나타내는 단어를 붙이는 게 좋습니다.

hwj0623 replied:
입력 형식을 잡아줘야 할 것 같은 느낌이라 기술하였는데 권장되지 않는 파라미터 레이블이었군요. 수정하겠습니다

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -5,6 +5,89 @@
+/// 2차원 사다리 문자열 배열 초기화 함수
+func initLadder(numOfPeople: Int, numOfLadders: Int) -> [[String]] {
+    let numOfColumn = numOfPeople * 2 - 1
+    let numOfRow = numOfLadders
```
스위프트에서는 줄여서 쓰는 것을 권장하지 않습니다. `numberOfColumn`, `numberOfRow` 그대로 쓰세요. 
값이 바뀌는 것도 아닌데 `numOfRow` 까지는 필요없지 않나요?

hwj0623 replied:
Column으로 쓰자니 Ladder도 Row로 바꿔야 2차원 배열의 행렬(row, column)의 가독성이 편해보인다는 생각이 있었습니다. 그런데 둘다 let 상수인거 보니 줄여도 될 것 같네요

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -5,6 +5,89 @@
+/// 2차원 사다리 문자열 배열의 print 함수
+func printLadder(ladderString2dArray :  [[String]]) -> Void {
+    for ( rowItems) in ladderString2dArray {
+        for ( columnItem) in rowItems {
```
사소한 거지만 스타일 기준에서는 공백, 띄어쓰기, 줄바꿈도 중요한 요소가 됩니다. 
구분하려는 어떤 의도가 있는게 아니라면 붙여서 쓰세요.

godrm reviewed (commented):
전체적인 구조나 흐름이 매우 좋습니다. 
다음 피드백을 확인해서 개선하고 넘어가면 될 것 같습니다.

hwj0623 committed:
PR 변경사항 반영
1) 입력함수(inputPairNumber)와 시작함수(startLadderGame)의 중복 디버그 메시지 수정
2) buildLadder, printLadder 함수의 파라미터레이블 변수명 수정
3) 줄바꿈 수정

새로 추가한 것
1) 유효성 체크 함수 - 문자열->숫자 변환가능여부, 숫자 유효범위 여부
2) enum으로 에러코드 관리

hwj0623 commented:
4-1 코드리뷰 개선사항 반영하였습니다. 
검토 후 PR merge 부탁드립니다.

godrm commented:
이번 커밋도 커밋로그가 나뉘어지지 않고 하나의 커밋이네요.
작은 변경 사항 작업 단위마다 커밋을 하고 한꺼번에 Push 하세요. 
커밋 중에서 일부를 선택하거나 되돌리거나 하는 작업이 가능하도록 작게 나눠야 합니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -9,21 +9,36 @@
+}
+/// 유효 숫자 범위 코드값
+enum ValidRangeCode: Int {
+    case validMarginalInputNumber = 2
```
enum 타입은 여러개 값이 있을때 주로 사용하는데 1개만 있다는 것은 어떤 의도가 있는건가요?

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -56,38 +72,90 @@ func  buildLadder (ladderInt2dArray : [[String]]) -> [[String]] {
+        ladders == ErrorCode.notANumber.rawValue) {
+        return ValidResultCode.invalid.rawValue
+    }
+    return ValidResultCode.valid.rawValue
```
정상 / 비정상을 구분하는 거라면 Bool 타입도 적합하지 않은가요?

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -56,38 +72,90 @@ func  buildLadder (ladderInt2dArray : [[String]]) -> [[String]] {
+/// 숫자 외 문자열 여부 체크 함수 - 정상 : 원래 값 정수형태로 반환, 예외 : 에러코드값 반환
+func checkInvalidCharacter(_ inputString: String) -> Int {
+    if inputString.isNumber() {
+        if let integerValue = Int (inputString){
```
`if let integerValue = Int (inputString)` 비교를 하면 inNumber 상위 비교를 안해도 문자가 포함되거나 숫자가 아닌 경우는 변환이 안되지 않을까요?

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -56,38 +72,90 @@ func  buildLadder (ladderInt2dArray : [[String]]) -> [[String]] {
+        printLadder(ladder2dMap: resultLadder)
     }else{
-        print("유효한 입력값을 넣어주세요 (인원, 높이 각각 2이상) \n. 양의 실수값 입력시 정수로 내림 처리됩니다.")
+        print("유효한 입력값을 넣어주세요 (인원, 높이 각각 2이상).\n양의 실수값 입력시 정수로 내림 처리됩니다.")
```
입력, 처리, 출력을 담당하는 하위 함수가 있는데 여기서는 print()를 직접 처리하고 있네요. 
상위 함수는 구체적인 동작을 하기 보다는 하위 함수를 호출하거나 조합해서 흐름을 조정하는 역할로 최소화하세요.
다음 단계를 진행할 때는 이 부분도 하위 함수로 분리해보세요.
